src/login/logind-session-device.c

```diff
@@ -367,9 +359,39 @@ int session_device_new(Session *s, dev_t dev, SessionDevice **out) {
         r = hashmap_put(s->devices, &sd->dev, sd);
         if (r < 0) {
                 r = -ENOMEM;
+                free(sd->node);
                 goto error;
         }
 
+        LIST_PREPEND(sd_by_device, sd->device->session_devices, sd);
+error:
+        return r;
+}
+
+static void session_device_deinit(SessionDevice *sd) {
+        LIST_REMOVE(sd_by_device, sd->device->session_devices, sd);
+        hashmap_remove(sd->session->devices, &sd->dev);
+        free(sd->node);
```

Comment:
i am really not feeling good about functions that "half-destroy" something but don't invalidate what they destroyed. Either functions should fully destroy things, or they should at least invalidate what they destroyed, i.e. "free(sd->node)" should be "sd->node = mfree(sd->node)", and so on. so that the function becomes idempotent.